src/Core/IdentityServer/BaseRequestValidator.cs

```diff
@@ -112,9 +115,19 @@ protected async Task ValidateAsync(T context, ValidatedTokenRequest request)
                 twoFactorToken = null;
             }
 
-            var device = await SaveDeviceAsync(user, request);
-            await BuildSuccessResultAsync(user, context, device, twoFactorRequest && twoFactorRemember);
-            return;
+            if (await RequiresSsoAuthAsync(user, request.GrantType)) // Returns true if can finish validation process
+            {
+                var device = await SaveDeviceAsync(user, request);
+                await BuildSuccessResultAsync(user, context, device, twoFactorRequest && twoFactorRemember);
+            }
+            else
+            {
+                SetSsoResult(context,
+                    new Dictionary<string, object>
```

Comment:
Formatting: Please move the `new Dictionary<TKey, TValue>` to the line above

Reply:
Will update!

Reply:
Updated.